externals/llvm-external-projects/torch-mlir-dialects/lib/Dialect/Tcp/IR/TcpOps.cpp

```diff
@@ -111,4 +111,52 @@ LogicalResult IsolatedGroupOp::verify() {
 
 OpFoldResult ConstOp::fold(FoldAdaptor) { return getValueAttr(); }
 
+LogicalResult ConcatOp::verify() {
+  auto outputTensorType = getResult().getType().cast<TensorType>();
+  int64_t axis = getAxis();
+  int64_t concatDimVal = outputTensorType.getShape()[axis];
+  // Accumulate concat dim, ignored if any of the inputs is dynamic.
+  int64_t concatDimAcc = 0;
+  // All tensors must have the smae dtype, rank and all non-concat dims must be
+  // the same.
+  for (auto type : getInputs().getTypes()) {
+    auto inputTensorType = type.cast<TensorType>();
+    if (inputTensorType.getElementType() != outputTensorType.getElementType())
+      return emitOpError() << "failed to verify tcp.concat with input and "
+                              "output operands rank mismatch";
+    if (outputTensorType.getRank() != inputTensorType.getRank())
+      return emitOpError() << "failed to verify tcp.concat with input and "
+                              "output operands rank mismatch";
+    for (int64_t dim = 0; dim < inputTensorType.getRank(); ++dim) {
+      if (dim == axis) {
+        concatDimAcc += inputTensorType.getShape()[dim];
+      } else {
+        if (inputTensorType.getShape()[dim] != outputTensorType.getShape()[dim])
+          emitOpError() << "failed to verify tcp.concat with non concat dim "
+                        << dim << ", having different values "
+                        << inputTensorType.getShape()[dim] << " "
+                        << outputTensorType.getShape()[dim];
+      }
+    }
+  }
+
+  // If concat dim is dynamic, at least one input must be dynamic.
+  if (ShapedType::isDynamic(concatDimVal)) {
+    if (!llvm::any_of(getInputs().getTypes(), [axis](Type type) {
+          return ShapedType::isDynamic(
+              type.cast<TensorType>().getShape()[axis]);
+        }))
+      return emitOpError() << "failed to verify tcp.concat with dynamic concat "
+                              "axis and static inputs";
+    else
+      return success();
+  }
+
```

Comment:
Very very minor nit (so optional to fix): IMO it will be cleaner to track this in `concatDimAcc` itself.  I.e. above instead of `concatDimAcc += inputTensorType.getShape()[dim];` do

```
if (ShapedType::isDynamic(inputTensorType.getShape()[dim]) || concatDimAcc == ShapedType::kDynamicSize)
  concatDimAcc = ShapedType::kDynamicSize;
else
  concatDimAcc += inputTensorType.getShape()[dim]
```

Then you just need to check `concatDimAcc != concatDimVal` here.